src/build/tenscript/pest_parser.rs

```diff
@@ -0,0 +1,184 @@
+            }
+            _ => unreachable!("build phase: {:?}", pair.as_rule()),
+        }
+    }
+    Ok(phase)
+}
+
+fn node(node_pair: Pair<Rule>) -> Result<TenscriptNode, ParseError> {
+    let pair = node_pair.into_inner().next().unwrap();
+    match pair.as_rule() {
+        Rule::face => {
+            let mut inner_pairs = pair.into_inner();
+            let face_name_string = inner_pairs.next().unwrap().as_str();
+            let face_name: FaceName = face_name_string[1..].try_into().unwrap();
+            let node = node(inner_pairs.next().unwrap()).unwrap();
+            Ok(TenscriptNode::Face {
+                face_name,
+                node: Box::new(node),
+            })
+        }
+        Rule::grow => {
+            let mut inner = pair.into_inner();
+            let forward_string = inner.next().unwrap().as_str();
+            let forward = match forward_string.parse() {
+                Ok(count) => { "X".repeat(count) }
+                Err(_) => { forward_string[1..forward_string.len() - 1].into() }
+            };
+            let scale_factor = scale(inner.next());
+            let post_growth_node = inner.next()
+                .map(|post_growth| Box::new(node(post_growth).unwrap()));
+            Ok(TenscriptNode::Grow {
+                forward,
+                scale_factor,
+                post_growth_node,
+            })
+        }
+        Rule::mark => {
+            let mark_name = pair.into_inner().next().unwrap().as_str();
+            Ok(TenscriptNode::Mark { mark_name: mark_name[1..].into() })
+        }
+        Rule::branch => {
+            Ok(TenscriptNode::Branch {
+                face_nodes: pair.into_inner()
+                    .map(|face_node| node(face_node).unwrap())
+                    .collect()
+            })
+        }
+        _ => unreachable!("node"),
+    }
+}
+
+fn scale(scale_pair: Option<Pair<Rule>>) -> f32 {
+    match scale_pair {
+        None => 1.0,
+        Some(scale_pair) => {
+            let scale_string = scale_pair.into_inner().next().unwrap().as_str();
+            scale_string.parse().unwrap()
+        }
+    }
+}
+
+pub fn parse(source: &str) -> Result<FabricPlan, ParseError> {
+    let mut pairs = PestParser::parse(Rule::fabric_plan, source)
+        .map_err(ParseError::PestError)?;
+    let plan_rule = pairs.next().unwrap();
+    fabric_plan(plan_rule)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::build::tenscript::bootstrap_fabric_plans;
+    use crate::build::tenscript::pest_parser::{parse, ParseError};
+
+    #[test]
+    fn parse_test() {
+        let plans = bootstrap_fabric_plans();
+        for (name, code) in plans.iter() {
+            match parse(code) {
+                Ok(plan) => {
+                    println!("[{name}] Good plan!");
+                    dbg!(plan);
+                }
+                Err(ParseError::PestError(error)) => panic!("[{name}] Error: {error}"),
+                Err(error) => panic!("[{name}] Error: {error:?}"),
```

Comment:
Alternatively, you could add an `impl Display for ParseError` and match on error type in there. All you'd have to put here is this
```suggestion
                Err(error) => panic!("[{name}] Error: {error}"),
```

Reply:
To conveniently display errors in other places